src/core/lib/surface/server.cc

```diff
@@ -1211,30 +1171,21 @@ void Server::ChannelData::InitTransport(RefCountedPtr<Server> server,
   grpc_transport_perform_op(transport, op);
 }
 
-Server::ChannelRegisteredMethod* Server::ChannelData::GetRegisteredMethod(
-    const grpc_slice& host, const grpc_slice& path) {
-  if (registered_methods_ == nullptr) return nullptr;
-  // TODO(ctiller): unify these two searches
-  // check for an exact match with host
-  uint32_t hash = MixHash32(grpc_slice_hash(host), grpc_slice_hash(path));
-  for (size_t i = 0; i <= registered_method_max_probes_; i++) {
-    ChannelRegisteredMethod* rm =
-        &(*registered_methods_)[(hash + i) % registered_methods_->size()];
-    if (rm->server_registered_method == nullptr) break;
-    if (!rm->has_host) continue;
-    if (rm->host != host) continue;
-    if (rm->method != path) continue;
-    return rm;
-  }
-  // check for a wildcard method definition (no host set)
-  hash = MixHash32(0, grpc_slice_hash(path));
-  for (size_t i = 0; i <= registered_method_max_probes_; i++) {
-    ChannelRegisteredMethod* rm =
-        &(*registered_methods_)[(hash + i) % registered_methods_->size()];
-    if (rm->server_registered_method == nullptr) break;
-    if (rm->has_host) continue;
-    if (rm->method != path) continue;
-    return rm;
+Server::RegisteredMethod* Server::ChannelData::GetRegisteredMethod(
+    const absl::string_view& host, const absl::string_view& path)
+    ABSL_LOCKS_EXCLUDED(registered_methods_mu_) {
+  MutexLock lock(&server_->registered_methods_mu_);
+  if (server_->registered_methods_.empty()) return nullptr;
+  // check for exact match with host
+  auto key = MixHash32(absl::HashOf(host), absl::HashOf(path));
```

Comment:
I don't think this is guaranteed unique.

We guarantee that a `pair<string,string>` of host & path is unique for a server, but the mix of the hash of the pair of strings is not.

We probably want to rephrase this so that registered_methods_ is a `flat_hash_map<pair<string,string>, ...>`

Reply:
Done.